src/compiler/checker.ts

```diff
@@ -16755,22 +16755,15 @@ namespace ts {
             return false;
         }
 
-        // Return the contextual type for a given expression node. During overload resolution, a contextual type may temporarily
-        // be "pushed" onto a node using the contextualType property.
-        function getApparentTypeOfContextualType(node: Expression): Type | undefined {
-            let contextualType = getContextualType(node);
-            contextualType = contextualType && mapType(contextualType, getApparentType);
-            if (!(contextualType && contextualType.flags & TypeFlags.Union && isObjectLiteralExpression(node))) {
-                return contextualType;
-            }
+        function discriminateContextualTypeByObjectMembers(node: ObjectLiteralExpression, contextualType: UnionType) {
             // Keep the below up-to-date with the work done within `isRelatedTo` by `findMatchingDiscriminantType`

```

Comment:
Since this applies to two functions now, enumerate the functions.

Actually, if you can extract the core of this algorithm and call it, that would be better. I'm not sure what you can do with that *hilarious* use of `break propLoop` in there. Maybe that could become a `return`?